File: src/cpp/id_selector.cpp
#include "faiss/impl/IDSelector.h"
#include "id_selector.hpp"


bool id_selector_is_member(const IDSelector *sel, int64_t id)
{
    return ((faiss::IDSelector*)sel)->is_member(id);
}

void id_selector_free(IDSelector *sel)
{
    delete (faiss::IDSelector*)sel;
}

IDSelectorRange *id_selector_range_new(int64_t imin, int64_t imax, bool assume_sorted)
{
    return (IDSelectorRange*)new faiss::IDSelectorRange(imin, imax, assume_sorted);
}

int64_t id_selector_range_imin(const IDSelectorRange *sel)
{
    return ((faiss::IDSelectorRange*)sel)->imin;
}

int64_t id_selector_range_imax(const IDSelectorRange *sel)
{
    return ((faiss::IDSelectorRange*)sel)->imax;
}

bool id_selector_range_assume_sorted(const IDSelectorRange *sel)
{
    return ((faiss::IDSelectorRange*)sel)->assume_sorted;
}

void id_selector_range_set_assume_sorted(IDSelectorRange *sel, bool assume_sorted)
{
    ((faiss::IDSelectorRange*)sel)->assume_sorted = assume_sorted;
}

IDSelectorArray *id_selector_array_new(int64_t n, const int64_t *ids)
{
    return (IDSelectorArray*)new faiss::IDSelectorArray(n, ids);
}

const int64_t *id_selector_array_ids(const IDSelectorArray *sel)
{
    return ((faiss::IDSelectorArray*)sel)->ids;
}

size_t id_selector_array_n(const IDSelectorArray *sel)
{
    return ((faiss::IDSelectorArray*)sel)->n;
}

IDSelectorBatch *id_selector_batch_new(int64_t n, const int64_t *ids)
{
    return (IDSelectorBatch*)new faiss::IDSelectorBatch(n, ids);
}


IDSelectorBitmap *id_selector_bitmap_new(int64_t n, const uint8_t *ids)
{
    return (IDSelectorBitmap*)new faiss::IDSelectorBitmap(n, ids);
}

IDSelectorNot *id_selector_not_new(const IDSelector *sel)
{
    return (IDSelectorNot*)new faiss::IDSelectorNot((faiss::IDSelector*)sel);
}

IDSelectorAll *id_selector_all_new()
{
    return (IDSelectorAll*)new faiss::IDSelectorAll();
}

IDSelectorAnd *id_selector_and_new(const IDSelector *lhs, const IDSelector *rhs)
{
    return (IDSelectorAnd*)new faiss::IDSelectorAnd((faiss::IDSelector*)lhs, (faiss::IDSelector*)rhs);
}

IDSelectorOr *id_selector_or_new(const IDSelector *lhs, const IDSelector *rhs)
{
    return (IDSelectorOr*)new faiss::IDSelectorOr((faiss::IDSelector*)lhs, (faiss::IDSelector*)rhs);
}

IDSelectorXOr *id_selector_xor_new(const IDSelector *lhs, const IDSelector *rhs)
{
    return (IDSelectorXOr*)new faiss::IDSelectorXOr((faiss::IDSelector*)lhs, (faiss::IDSelector*)rhs);
}

File: src/id_selector.rs
use std::ptr;

use faiss_next_sys::{
    self, FaissIDSelector, FaissIDSelectorAnd, FaissIDSelectorBatch, FaissIDSelectorNot,
    FaissIDSelectorOr, FaissIDSelectorRange, FaissIDSelectorXOr,
};

use crate::error::{check_return_code, Result};
use crate::idx::Idx;

pub trait IDSelector {
    fn as_ptr(&self) -> *const FaissIDSelector;
}

pub struct IDSelectorRange {
    ptr: *mut FaissIDSelectorRange,
}

impl IDSelectorRange {
    pub fn new(imin: i64, imax: i64) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorRange = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorRange_new(
                &mut ptr, imin, imax,
            ))?;
            Ok(Self { ptr })
        }
    }

    pub fn imin(&self) -> i64 {
        unsafe { faiss_next_sys::faiss_IDSelectorRange_imin(self.ptr) }
    }

    pub fn imax(&self) -> i64 {
        unsafe { faiss_next_sys::faiss_IDSelectorRange_imax(self.ptr) }
    }
}

impl IDSelector for IDSelectorRange {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorRange {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelectorRange_free(self.ptr);
            }
        }
    }
}

pub struct IDSelectorBatch {
    ptr: *mut FaissIDSelectorBatch,
}

impl IDSelectorBatch {
    pub fn new(indices: &[i64]) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorBatch = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorBatch_new(
                &mut ptr,
                indices.len(),
                indices.as_ptr(),
            ))?;
            Ok(Self { ptr })
        }
    }

    pub fn from_idx(indices: &[Idx]) -> Result<Self> {
        let ids: Vec<i64> = indices.iter().map(|&id| id.as_repr()).collect();
        Self::new(&ids)
    }
}

impl IDSelector for IDSelectorBatch {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorBatch {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelector_free(self.ptr as *mut FaissIDSelector);
            }
        }
    }
}

pub struct IDSelectorNot {
    ptr: *mut FaissIDSelectorNot,
}

impl IDSelectorNot {
    pub fn new<S: IDSelector>(sel: &S) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorNot = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorNot_new(
                &mut ptr,
                sel.as_ptr(),
            ))?;
            Ok(Self { ptr })
        }
    }
}

impl IDSelector for IDSelectorNot {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorNot {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelector_free(self.ptr as *mut FaissIDSelector);
            }
        }
    }
}

pub struct IDSelectorAnd {
    ptr: *mut FaissIDSelectorAnd,
}

impl IDSelectorAnd {
    pub fn new<L: IDSelector, R: IDSelector>(lhs: &L, rhs: &R) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorAnd = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorAnd_new(
                &mut ptr,
                lhs.as_ptr(),
                rhs.as_ptr(),
            ))?;
            Ok(Self { ptr })
        }
    }
}

impl IDSelector for IDSelectorAnd {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorAnd {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelector_free(self.ptr as *mut FaissIDSelector);
            }
        }
    }
}

pub struct IDSelectorOr {
    ptr: *mut FaissIDSelectorOr,
}

impl IDSelectorOr {
    pub fn new<L: IDSelector, R: IDSelector>(lhs: &L, rhs: &R) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorOr = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorOr_new(
                &mut ptr,
                lhs.as_ptr(),
                rhs.as_ptr(),
            ))?;
            Ok(Self { ptr })
        }
    }
}

impl IDSelector for IDSelectorOr {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorOr {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelector_free(self.ptr as *mut FaissIDSelector);
            }
        }
    }
}

pub struct IDSelectorXOr {
    ptr: *mut FaissIDSelectorXOr,
}

impl IDSelectorXOr {
    pub fn new<L: IDSelector, R: IDSelector>(lhs: &L, rhs: &R) -> Result<Self> {
        unsafe {
            let mut ptr: *mut FaissIDSelectorXOr = ptr::null_mut();
            check_return_code(faiss_next_sys::faiss_IDSelectorXOr_new(
                &mut ptr,
                lhs.as_ptr(),
                rhs.as_ptr(),
            ))?;
            Ok(Self { ptr })
        }
    }
}

impl IDSelector for IDSelectorXOr {
    fn as_ptr(&self) -> *const FaissIDSelector {
        self.ptr as *const FaissIDSelector
    }
}

impl Drop for IDSelectorXOr {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                faiss_next_sys::faiss_IDSelector_free(self.ptr as *mut FaissIDSelector);
            }
        }
    }
}
